types/universal-router/index.d.ts

```diff
@@ -1,68 +1,118 @@
  */
-export interface Route<C, R> {
-  path: string;
-  action(ctx: ActionContext & C, params: Params): R | Promise<R> | void;
-  children?: Routes<C, R>;
+export interface Route<C extends Context = any, R = any> {
+    path?: string | RegExp | Array<string | RegExp>;
+    /**
+     * unique string
+     */
+    name?: string;
+    /**
+     * automatically filled by the router
+     */
+    parent?: Route | null;
+    /**
+     * array of route objects
+     */
+    children?: Routes<C, R> | null;
+    /**
+     * action method should return anything except `null` or `undefined` to be resolved by router
+     * otherwise router will throw `Page not found` error if all matched routes returned nothing
+     */
+    action?: (context: ActionContext<C, R> & C, params: Params) => R | Promise<R> | void;
 }
 
 /**
  * Routes in an array of type Route.
  * @template C User context that is made union with ActionContext.
- * @template R Result that every action function resolves to. If the action
- *  returns a Promise, R can be the type the Promise resolves to.
+ * @template R Result that every action function resolves to. If the action returns a Promise, R can be the type the Promise resolves to.
  */
-export type Routes<C, R> = Array<Route<C, R>>;
+export type Routes<C extends Context = Context, R = any> = Array<Route<C, R>>;
+
+export interface Options<C extends Context = Context, R = any> {
+    context?: C;
+    baseUrl?: string;
+    resolveRoute?: (context: ActionContext<C, R> & C, params: Params) => any;
+    errorHandler?: (error: Error & { context: C, code: number }) => any;
+}
 
 /**
- * Resolve function that is given routes and a path or context object.
- * Returns a Promise that resolves to result of the action function of the
- * matched route.
- *
- * @template C User context that is made union with Context.
- * @template R Result that every action function resolves to. If the action
- *  returns a Promise, R can be the type the Promise resolves to.
- *
- * @param routes - Single route or array of routes.
- * @param pathOrContext - path to resolve or
- *  context object that contains the path along with other data.
- * @return - Result of matched action function wrapped in a Promsie.
+ * @see https://github.com/kriasoft/universal-router/blob/v6.0.0/docs/api.md
  */
-export function resolve<C, R>(routes: Routes<C, R> | Route<C, R>, pathOrContext: string | String | Context & C): Promise<R>;
+export default class UniversalRouter<C extends Context = Context, R = any> {
+    constructor(routes: Routes<C, R> | Route<C, R>, options?: Options<C>)
+    /**
+     * Traverses the list of routes in the order they are defined until it finds the first route that matches provided URL path string
+     * and whose action function returns anything other than null or undefined.
+     */
+    resolve(pathnameOrContext: string | PathnameContext): Promise<R>;
+    /**
+     * `path-to-regexp` function
+     */
+    static pathToRegexp: any;
```

Comment:
This property could also be typed with imported `pathToRegex` typings as I mentioned above